File: packages/python/src/sererr/_capture.py
"""Walk a Python exception chain into a list of :class:`CapturedError`.

Python's exception-chain semantics (PEP 3134):

- ``__cause__`` is set by ``raise X from Y`` (explicit).
- ``__context__`` is set automatically when ``raise`` happens inside an
  ``except`` handler (implicit).
- ``__suppress_context__`` is set by ``raise X from None`` — meaning the
  user explicitly silenced the implicit context.

The standard precedence (matching how ``traceback.format_exception``
walks the chain) is: prefer ``__cause__``; otherwise use
``__context__`` unless ``__suppress_context__`` is set.

Frame ordering: :class:`traceback.StackSummary` is oldest-first (the
outermost caller first, the actual ``raise`` site last). The sererr
contract requires **most-recent-first**, so we reverse on encode.
"""

from __future__ import annotations

import traceback

from sererr.types import CapturedError, ExceptionMechanism, StackFrame


def _next_in_chain(exc: BaseException) -> BaseException | None:
    """Walk one step deeper into the cause/context chain.

    Returns ``None`` when the chain terminates. ``__cause__`` wins over
    ``__context__``; ``__suppress_context__`` (set by
    ``raise ... from None``) terminates the chain entirely if there is
    no explicit cause.
    """
    if exc.__cause__ is not None:
        return exc.__cause__
    if exc.__suppress_context__:
        return None
    return exc.__context__


def _frames_for(exc: BaseException) -> list[StackFrame]:
    """Extract frames from an exception's traceback as most-recent-first
    :class:`StackFrame` instances.

    ``traceback.TracebackException.from_exception(exc).stack`` returns a
    :class:`traceback.StackSummary` ordered oldest-first; the sererr
    contract is most-recent-first, so we reverse.
    """
    tbe = traceback.TracebackException.from_exception(exc)
    frames: list[StackFrame] = []
    # StackSummary is iterable of FrameSummary; oldest-first.
    for fs in tbe.stack:
        frames.append(
            StackFrame(
                function=fs.name or "",
                file=fs.filename or "",
                line=fs.lineno or 0,
                context_line=(fs.line or "") if fs.line is not None else "",
            )
        )
    frames.reverse()
    return frames


def capture(
    exc: BaseException,
    type_name: str,
    release: str,
    server_name: str,
) -> list[CapturedError]:
    """Walk a Python exception chain into a sererr capture.

    The returned list is **most-causal-first** — the originating caught
    error is the last element. Each entry carries its own frames
    (captured from that exception's own ``__traceback__``).

    Mechanism IDs are stamped sequentially: ``exception_id = index``,
    ``parent_id = max(0, index - 1)``. The root cause has
    ``parent_id == 0`` (itself).

    ``type_name`` labels the *originating caught* error (the LAST entry
    in the returned chain). Inner entries derive their type from
    ``type(e).__name__`` since the live Python exception object carries
    its own type information.
    """
    # Walk outward → in: start at the caught exception, follow
    # __cause__/__context__ inward. Collect outer-first, then reverse so
    # the originating caught error is LAST (most-causal-first order).
    chain_outer_first: list[CapturedError] = []
    current: BaseException | None = exc
    is_outermost = True
    while current is not None:
        # Outermost entry uses caller-supplied type_name (Rust convention);
        # deeper entries use the live exception type.
        t = type_name if is_outermost else type(current).__name__
        chain_outer_first.append(
            CapturedError(
                type=t,
                message=str(current),
                frames=_frames_for(current),
                mechanism=ExceptionMechanism(type="generic", handled=True),
                release=release,
                server_name=server_name,
            )
        )
        is_outermost = False
        current = _next_in_chain(current)

    # Reverse to most-causal-first; the originating caught error is last.
    chain = list(reversed(chain_outer_first))

    for i, entry in enumerate(chain):
        if entry.mechanism is not None:
            entry.mechanism.exception_id = i
            entry.mechanism.parent_id = 0 if i == 0 else i - 1
    return chain

File: packages/csharp/src/Capture.cs
using System.Diagnostics;
using System.Reflection;

namespace Sererr;

/// <summary>
/// Top-level entry point: walks an <see cref="Exception"/> into a
/// <see cref="CapturedError"/> chain.
///
/// The returned list is <b>most-causal-first</b> — the originating
/// caught error is the LAST element. Each entry carries the same
/// frames captured from the supplied exception; chain position is
/// identified by <c>mechanism.exception_id</c>.
/// </summary>
public static class Capture
{
    /// <summary>
    /// Walk an <see cref="Exception"/> into a sererr capture chain.
    /// </summary>
    /// <param name="exception">The caught exception.</param>
    /// <param name="typeName">User-supplied label for the outermost
    /// (caught) exception; subsequent entries use their runtime type's
    /// fully-qualified name.</param>
    /// <param name="release">Build identifier (semver / git SHA).</param>
    /// <param name="serverName">Producing host / pod name.</param>
    public static IReadOnlyList<CapturedError> Build(
        Exception exception,
        string typeName,
        string release,
        string serverName)
    {
        // Capture frames from the leaf exception's stack trace. .NET's
        // StackTrace(exception, fNeedFileInfo: true) returns frames
        // most-recent-call-first; no reversal needed.
        var frames = ExtractFrames(exception);

        // Walk the exception tree into a flat most-causal-first list.
        // For each entry we record its (post-reverse) parent index in
        // the proto sense:
        //   - linear chain: parent = i-1 (the inner cause)
        //   - aggregate siblings: parent = the aggregate's index
        //
        // We accumulate raw entries first, then reverse to get
        // most-causal-first ordering, and stamp parent_id in a second
        // pass using the simple `i-1` formula except where the original
        // walker recorded a sibling-of-aggregate relationship.
        var raw = new List<RawEntry>();
        WalkOutermostFirst(exception, typeName, raw, aggregateParent: -1);

        // raw is outermost-first; reverse to most-causal-first.
        raw.Reverse();
        var n = raw.Count;
        // Remap the recorded aggregateParent (a pre-reverse index) to a
        // post-reverse index. -1 stays -1.
        for (var i = 0; i < n; i++)
        {
            var pre = raw[i].AggregateParentPreReverse;
            raw[i].AggregateParentPostReverse = pre < 0 ? -1 : (n - 1 - pre);
        }

        var chain = new List<CapturedError>(n);
        for (var i = 0; i < n; i++)
        {
            var entry = raw[i];
            // parent_id rules in most-causal-first order:
            //   1. Aggregate sibling → parent = aggregate's index.
            //   2. Aggregate itself → parent = 0 (no single causal parent
            //      across multiple children).
            //   3. Linear chain → parent = i-1 (the inner cause we wrap),
            //      clamped to 0 at the root.
            uint parentId;
            if (entry.AggregateParentPostReverse >= 0)
            {
                parentId = (uint)entry.AggregateParentPostReverse;
            }
            else if (entry.IsExceptionGroup)
            {
                parentId = 0u;
            }
            else
            {
                parentId = i == 0 ? 0u : (uint)(i - 1);
            }

            chain.Add(new CapturedError
            {
                Type = entry.TypeName,
                Message = entry.Message,
                Frames = frames,
                Mechanism = new ExceptionMechanism
                {
                    Type = "generic",
                    Handled = true,
                    ExceptionId = (uint)i,
                    ParentId = parentId,
                    IsExceptionGroup = entry.IsExceptionGroup,
                },
                Release = release,
                ServerName = serverName,
            });
        }

        return chain;
    }

    // ---- Internal walker --------------------------------------------------

    private sealed class RawEntry
    {
        public required string TypeName { get; init; }
        public required string Message { get; init; }
        public bool IsExceptionGroup { get; init; }
        /// <summary>
        /// Pre-reverse index of the aggregate parent (if this entry is
        /// a sibling under an AggregateException); -1 otherwise.
        /// </summary>
        public int AggregateParentPreReverse { get; init; } = -1;
        public int AggregateParentPostReverse { get; set; } = -1;
    }

    /// <summary>
    /// Walk <paramref name="ex"/> into <paramref name="nodes"/>,
    /// outermost first. Aggregate children record their aggregate's
    /// pre-reverse index so we can stamp parent_id later; linear chain
    /// links are reconstructed positionally.
    /// </summary>
    private static void WalkOutermostFirst(
        Exception ex,
        string? typeName,
        List<RawEntry> nodes,
        int aggregateParent)
    {
        var isGroup = ex is AggregateException;
        nodes.Add(new RawEntry
        {
            TypeName = typeName ?? ex.GetType().FullName ?? ex.GetType().Name,
            Message = ex.Message,
            IsExceptionGroup = isGroup,
            AggregateParentPreReverse = aggregateParent,
        });
        var myIdx = nodes.Count - 1;

        if (ex is AggregateException agg)
        {
            // Aggregate siblings record THIS aggregate as their parent.
            // The aggregate itself uses default linear parenting (or
            // -1 if it has none).
            foreach (var inner in agg.InnerExceptions)
            {
                WalkOutermostFirst(inner, typeName: null, nodes, aggregateParent: myIdx);
            }
        }
        else if (ex.InnerException is { } inner)
        {
            // Linear chain — no aggregate context.
            WalkOutermostFirst(inner, typeName: null, nodes, aggregateParent: -1);
        }
    }

    // ---- Frame extraction --------------------------------------------------

    private static IReadOnlyList<StackFrame> ExtractFrames(Exception exception)
    {
        // Ben.Demystifier cleans up async state-machine MoveNext frames
        // and similar compiler detritus. The Demystify() call mutates
        // exception state that affects later ToString() rendering and
        // can also trim the StackTrace output. Tolerate failure — it
        // must never break capture.
        try
        {
            exception.Demystify();
        }
        catch
        {
        }

        var trace = new StackTrace(exception, fNeedFileInfo: true);
        var raw = trace.GetFrames();
        if (raw is null || raw.Length == 0)
        {
            return Array.Empty<StackFrame>();
        }

        var frames = new List<StackFrame>(raw.Length);
        foreach (var sf in raw)
        {
            if (sf is null) continue;
            var method = sf.GetMethod();
            var function = FormatMethod(method);
            var module = method?.DeclaringType?.FullName ?? "";
            var package = method?.DeclaringType?.Assembly.GetName().Name ?? "";
            var file = sf.GetFileName() ?? "";
            var lineNumber = sf.GetFileLineNumber();
            var line = lineNumber > 0 ? (uint)lineNumber : 0u;

            frames.Add(new StackFrame
            {
                Function = function,
                Module = module,
                Package = package,
                File = file,
                AbsPath = file,
                Line = line,
                InApp = IsAppFrame(module, package),
            });
        }
        return frames;
    }

    private static string FormatMethod(MethodBase? method)
    {
        if (method is null) return "<unknown>";
        var declaring = method.DeclaringType;
        var typeName = declaring?.FullName ?? declaring?.Name ?? "";
        return string.IsNullOrEmpty(typeName)
            ? method.Name
            : $"{typeName}.{method.Name}";
    }

    /// <summary>
    /// Default <c>in_app</c> heuristic. Frames whose module/package
    /// looks like .NET runtime / framework code are flagged false.
    /// </summary>
    private static bool IsAppFrame(string module, string package)
    {
        if (string.IsNullOrEmpty(module) && string.IsNullOrEmpty(package))
        {
            return false;
        }
        ReadOnlySpan<string> nonAppPrefixes = new[]
        {
            "System.",
            "Microsoft.",
            "Internal.",
        };
        foreach (var prefix in nonAppPrefixes)
        {
            if (module.StartsWith(prefix, StringComparison.Ordinal)) return false;
            var trimmed = prefix.TrimEnd('.');
            if (package.StartsWith(trimmed, StringComparison.Ordinal)) return false;
        }
        return true;
    }
}
